File: examples/python/identify_the_customer.py
# Step 1 — Identify the customer (Python / FastAPI)
#
# Create the MindCloud end user once, store the id on your own user row, then
# mint a short-lived token whenever that customer opens your integrations page.
import os

import httpx

API = "https://connect.mindcloud.co"
HEADERS = {"Authorization": f"Bearer {os.environ['MINDCLOUD_API_KEY']}"}


# POST /v1/users does NOT deduplicate — calling it twice creates two end users.
async def create_end_user(user) -> str:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{API}/v1/users",
            headers=HEADERS,
            json={"externalId": user.id, "name": user.name, "email": user.email},
        )

    return response.json()["data"]["userId"]


@app.post("/api/embedded-token")
async def embedded_token(user=Depends(current_user)):
    end_user_id = user.mindcloud_end_user_id

    if not end_user_id:
        end_user_id = await create_end_user(user)
        await db.users.update(user.id, mindcloud_end_user_id=end_user_id)

    async with httpx.AsyncClient() as client:
        response = await client.get(f"{API}/v1/users/{end_user_id}/token", headers=HEADERS)

    return {"token": response.json()["token"]}

File: examples/node/identify-the-customer.js
// Step 1 — Identify the customer (Node.js / Express)
//
// Create the MindCloud end user once, store the id on your own user row, then
// mint a short-lived token whenever that customer opens your integrations page.
// Your API key stays on the server; the browser only ever holds the token.
const API = 'https://connect.mindcloud.co';

const headers = {
  Authorization: `Bearer ${process.env.MINDCLOUD_API_KEY}`,
  'Content-Type': 'application/json'
};

// POST /v1/users does NOT deduplicate — calling it twice creates two end users.
const createEndUser = async (user) => {
  const response = await fetch(`${API}/v1/users`, {
    method: 'POST',
    headers,
    body: JSON.stringify({ externalId: user.id, name: user.name, email: user.email })
  });

  const body = await response.json();
  return body.data.userId;
};

router.post('/api/embedded-token', async (req, res) => {
  const user = req.session.user;

  let endUserId = user.mindcloudEndUserId;
  if (!endUserId) {
    endUserId = await createEndUser(user);
    await db.users.update(user.id, { mindcloudEndUserId: endUserId });
  }

  const response = await fetch(`${API}/v1/users/${endUserId}/token`, { headers });
  const { token } = await response.json();

  res.json({ token });
});
